### PR TITLE
fix(platform): Fix Google Drive auto-credentials handling across the platform

### PR DESCRIPTION
**Pull request by ntindle**
Google Drive file picker blocks embed a `_credentials_id` in the file field data, but the platform had no way to distinguish these "auto-credentials" from regular user-provided credentials. This caused auto-credentials to appear in UI credential forms, CoPilot "missing credentials" prompts, and library preset guards — confusing users. Additionally, forking an agent didn't clear the embedded credential references, so cloned agents could silently use the original author's Google credentials. Finally, Firefox's strict referrer policy was blocking Google API script loads.

### Changes 🏗️

**Backend — Data model (`model.py`, `block.py`, `graph.py`)**
- `CredentialsFieldInfo` gains `is_auto_credential` and `input_field_name` fields to tag auto-credentials
- `BlockSchema.get_auto_credentials_fields()` returns auto-credential metadata for execution
- `Graph.regular_credentials_inputs` / `auto_credentials_inputs` properties filter aggregated credentials
- `_reassign_ids()` clears `_credentials_id` on fork to prevent cross-user credential leakage
- `combine()` and `discriminate()` propagate `is_auto_credential` and `input_field_name`

**Backend — Executor (`manager.py`, `utils.py`)**
- `_acquire_auto_credentials()` extracted as standalone async helper to resolve embedded `_credentials_id` at execution time with Redis lock management
- `_validate_node_input_credentials()` validates auto-credential ownership before execution
- `make_node_credentials_input_map()` uses `regular_credentials_inputs` only

**Backend — CoPilot & Library (`chat/tools/utils.py`, `library/db.py`)**
- `build_missing_credentials_from_graph()` and `match_user_credentials_to_graph()` use `regular_credentials_inputs` so auto-credentials aren't reported as "missing"
- Library preset guard uses `regular_credentials_inputs`

**Frontend (`GoogleDrivePicker/helpers.ts`)**
- Adds `referrerPolicy: "no-referrer-when-downgrade"` to both Google API script loads (gapi + picker) to fix Firefox blocking

**Known deferred:** `run_block.py` (CoPilot `RunBlockTool`) still includes auto-credentials in credential matching — tracked as SECRT-1911

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] `manager_auto_credentials_test.py` — 6 tests covering credential acquisition, error cases, chained data
  - [x] `utils_test.py` — 8 tests covering auto-credential validation, fork clearing, mixed credential graphs
  - [x] `graph_test.py` — 5 tests covering `regular_credentials_inputs`, `auto_credentials_inputs`, schema exclusion, `combine()`, `discriminate()` propagation
  - [x] `chat/tools/utils_test.py` — 2 tests covering CoPilot credential filtering
  - [x] All 916 existing tests pass (84 skipped, 0 failures)
  - [x] All pre-commit hooks pass (linting, formatting, typechecking)

#### For configuration changes:
- [x] `.env.default` is updated or already compatible with my changes
- [x] `docker-compose.yml` is updated or already compatible with my changes
- [x] I have included a list of my configuration changes in the PR description (under **Changes**) — N/A, no configuration changes

---

> [!NOTE]
> **Medium Risk**
> Touches credential schema aggregation and executor credential acquisition/validation paths; while well-tested, mistakes could block runs or incorrectly prompt for credentials across many agents.
> 
> **Overview**
> Fixes Google Drive *auto-credentials* end-to-end by tagging them in `CredentialsFieldInfo` (`is_auto_credential`, `input_field_name`) and splitting graph credential aggregation into `regular_credentials_inputs` vs `auto_credentials_inputs`, so only regular creds appear in graph schemas, CoPilot “missing credentials” prompts, credential mapping, and preset eligibility checks.
> 
> Strengthens execution-time handling for file-embedded `_credentials_id`: forks now strip `_credentials_id` to prevent cross-user credential reuse, executor validation checks auto-credential ownership/missing auth, and execution resolves auto-creds via a new `_acquire_auto_credentials()` helper with clearer error messages; adds extensive regression tests and sets `referrerPolicy` on Google script loads to avoid Firefox blocking.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c42aab3482066f882a8f6983131c86d0aa9c7a4f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: autogpt_platform/backend/backend/data/model.py [hunk 1]
```diff
@@ -654,6 +656,9 @@ def combine(
                 + "_credentials"
             )
 
+            # Propagate is_auto_credential from the combined field.
+            # All fields in a group should share the same is_auto_credential
+            # value since auto and regular credentials serve different purposes.
             result[group_key] = (
                 CredentialsFieldInfo[CP, CT](
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/data/model.py:690]: This is intentional. Google OAuth credentials from the file picker ARE Google OAuth credentials — they use the same provider and type. Separating them in the grouping key would force users to authenticate twice (once for the file picker, once for the regular credential), which is worse UX. The Google OAuth scopes are additive, so a single credential works for both paths.

File: autogpt_platform/backend/backend/executor/manager.py [hunk 2]
```diff
@@ -271,41 +346,14 @@ async def execute_node(
...
-                file_name = field_data.get("name", "selected file")
-                raise ValueError(
-                    f"Authentication missing for '{file_name}' in field '{field_name}'. "
-                    f"Please re-select the file to authenticate with {provider.capitalize()}."
-                )
+    auto_extra_kwargs, auto_locks = await _acquire_auto_credentials(
+        input_model=input_model,
+        input_data=input_data,
+        creds_manager=creds_manager,
+        user_id=user_id,
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/manager.py:391]: Acknowledged — there's already a comment at lines 191-194 documenting this. Currently no block has multiple auto-credential fields, and locks auto-expire via Redis TTL. If this becomes a real scenario we'll add a try/except to release partial locks.

File: autogpt_platform/backend/backend/executor/utils.py [hunk 3]
```diff
@@ -339,6 +340,31 @@ async def _validate_node_input_credentials(
...
+                        creds_store = get_integration_credentials_store()
+                        creds = await creds_store.get_creds_by_id(user_id, cred_id)
+                        if not creds:
+                            has_missing_credentials = True
+                            credential_errors[node.id][field_name] = (
+                                "The saved Google credentials are not available "
+                                "for your account. Please re-select the file to "
+                                "authenticate with your own Google account."
+                            )
+
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/utils.py:384]: Considered this, but the edge cases (empty dict, dict without `_credentials_id`, etc.) would require either a frontend bug or manual API tampering — the Google Drive picker sends a specific format. `_acquire_auto_credentials()` in manager.py already catches all these cases at runtime with clear error messages, so the practical benefit of duplicating that validation here is minimal. Skipping to avoid over-engineering.

File: autogpt_platform/backend/backend/data/model.py [hunk 4]
```diff
@@ -654,6 +656,9 @@ def combine(
                 + "_credentials"
             )
 
+            # Propagate is_auto_credential from the combined field.
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/data/model.py:686]: False positive — same reasoning as above. Google OAuth credentials from the file picker ARE Google OAuth credentials (same provider, same type). Separating them in the grouping key would force double authentication. Google OAuth scopes are additive, so a single credential handles both regular and auto-credential paths.

File: autogpt_platform/backend/backend/executor/manager.py [hunk 5]
```diff
@@ -172,6 +172,81 @@ def execute_graph(
...
+        else:
+            # field_data is None/empty but key IS in input_data = user didn't select
+            provider = info.get("config", {}).get("provider", "external service")
+            raise ValueError(
+                f"No file selected for '{field_name}'. "
+                f"Please select a file to provide "
+                f"{provider.capitalize()} authentication."
+            )
+
+    return extra_exec_kwargs, locks
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/manager.py:244]: Acknowledged — documented with a comment at lines 191-194. Currently no block has multiple auto-credential fields, and Redis TTL provides a safety net. Will add proper cleanup if multi-field auto-credentials become a real scenario.

File: autogpt_platform/backend/backend/data/graph.py [hunk 6]
```diff
@@ -487,6 +487,28 @@ def aggregate_credentials_inputs(
...
+    @property
+    def auto_credentials_inputs(
+        self,
+    ) -> dict[str, tuple[CredentialsFieldInfo, set[tuple[str, str]]]]:
+        """Credentials embedded in file fields (_credentials_id), resolved at execution time."""
+        return {
+            k: v
+            for k, v in self.aggregate_credentials_inputs().items()
+            if v[0].is_auto_credential
+        }
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/data/graph.py:638]: Intentional — this property is the counterpart to `regular_credentials_inputs` and will be used by the CoPilot RunBlockTool auto-credentials filtering work tracked in SECRT-1911. Keeping it to avoid re-implementing when that work lands.

File: autogpt_platform/backend/backend/executor/utils.py [hunk 7]
```diff
@@ -339,6 +340,31 @@ async def _validate_node_input_credentials(
...
+                if nodes_input_masks and node.id in nodes_input_masks:
+                    field_value = nodes_input_masks[node.id].get(
+                        field_name, field_value
+                    )
+
+                if field_value and isinstance(field_value, dict):
+                    cred_id = field_value.get("_credentials_id")
+                    if cred_id and isinstance(cred_id, str):
+                        creds_store = get_integration_credentials_store()
+                        creds = await creds_store.get_creds_by_id(user_id, cred_id)
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/utils.py]: Good catch — fixed in c5abc01. Added try/except matching the regular credentials path pattern.

File: autogpt_platform/backend/backend/executor/utils.py [hunk 8]
```diff
@@ -339,6 +340,31 @@ async def _validate_node_input_credentials(
...
+                        if not creds:
+                            has_missing_credentials = True
+                            credential_errors[node.id][field_name] = (
+                                "The saved Google credentials are not available "
+                                "for your account. Please re-select the file to "
+                                "authenticate with your own Google account."
+                            )
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/utils.py:383]: Good catch — fixed in c5abc01. Added try/except matching the regular credentials path pattern.

File: autogpt_platform/backend/backend/data/graph.py [hunk 9]
```diff
@@ -654,6 +675,16 @@ def _reassign_ids(
...
 
+        # Clear auto-credentials references (e.g., _credentials_id in
+        # GoogleDriveFile fields) so the new user must re-authenticate
+        # with their own account
+        for node in graph.nodes:
+            if not node.input_default:
+                continue
+            for key, value in node.input_default.items():
+                if isinstance(value, dict) and "_credentials_id" in value:
+                    value["_credentials_id"] = None
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/data/graph.py]: Good catch — fixed in f4f81bc. Changed `_reassign_ids` to `del value["_credentials_id"]` instead of setting it to None. This makes the three states unambiguous: key present with value = user-selected, `None` = chained upstream data, absent = forked/needs re-auth. Also added pre-run validation for the missing key case so the error surfaces at validation time, not as a confusing TypeError at runtime.

File: autogpt_platform/backend/backend/executor/manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/executor/manager.py:244]: Same issue flagged by Copilot and Cursor — already documented with a comment at lines 191-194. Currently no block has multiple auto-credential fields, and Redis TTL provides a safety net. Will add a try/finally to release partial locks if multi-field auto-credentials become a real scenario.

File: autogpt_platform/backend/backend/data/graph_test.py [hunk 10]
```diff
@@ -463,3 +463,328 @@ def test_node_credentials_optional_with_other_metadata():
     assert node.credentials_optional is True
     assert node.metadata["position"] == {"x": 100, "y": 200}
     assert node.metadata["customized_name"] == "My Custom Node"
+
+
+# ============================================================================
+# Tests for _reassign_ids credential clearing (Fix 3: SECRT-1772)
+def test_combine_preserves_is_auto_credential_flag():
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/backend/backend/data/graph_test.py:469]: Good catch — fixed in 90b3b5b. Moved the section header to above the actual `_reassign_ids` tests and relabeled the `combine()` section.

File: autogpt_platform/backend/backend/executor/manager.py [hunk 11]
```diff
@@ -172,6 +172,81 @@ def execute_graph(
 T = TypeVar("T")
 
 
+async def _acquire_auto_credentials(
```
Comment (majdyz) [inline autogpt_platform/backend/backend/executor/manager.py:172]: can this be moved to utils.py altogether? I see similar code in utils.py, are those replicated logic?

**Review by majdyz — DISMISSED** [review]
@Otto-AGPT can you check my comments and other ai pr review comments and whether it make sense to address? if not then this is good to merge

File: autogpt_platform/backend/backend/executor/manager.py (hunk 11 above)
Comment (Otto-AGPT) [reply autogpt_platform/backend/backend/executor/manager.py:172]: Good catch! Looking at this:

**In `manager.py` (`_acquire_auto_credentials`):**
- Iterates over `get_auto_credentials_fields()`
- Checks for `_credentials_id` in field data
- **Acquires credentials + locks** via `creds_manager.acquire()`
- Returns credentials for actual execution use

**In `utils.py` (`_validate_node_input_credentials`):**
- Also iterates over `get_auto_credentials_fields()`
- Also checks for `_credentials_id` in field data
- **Only validates** credentials exist via `get_creds_by_id()`
- No locking, just returns validation errors

**Analysis:**
The field traversal logic (checking `_credentials_id`, extracting provider info for error messages) is duplicated. However, the operations differ:
- Validation: read-only check, no locks
- Acquisition: locking + returns actual credentials

**Suggestion:** Could extract a shared helper like `extract_auto_credential_info(field_name, info, input_data) -> (cred_id, provider, file_name) | ValidationError` that both functions use. The validation vs acquisition logic stays separate but the parsing/error-message-building becomes DRY.

@ntindle thoughts on this? Happy to help refactor if useful.

Comment (Otto-AGPT) [reply autogpt_platform/backend/backend/executor/manager.py:172]: Done! Created [PR #12016](https://github.com/Significant-Gravitas/AutoGPT/pull/12016) that:

1. Adds `AutoCredentialFieldInfo` dataclass and `parse_auto_credential_field()` helper to `executor/utils.py`
2. Updates `_acquire_auto_credentials` in `manager.py` to use the shared helper
3. Updates `_validate_node_input_credentials` in `utils.py` to use the shared helper

This keeps the distinct operations (acquire with locking vs validate without) separate while consolidating the shared parsing logic. Makes `manager.py` less cluttered as requested.

The PR is based on this branch so it can be merged after #12004.

**Review by autogpt-reviewer — COMMENTED** [review]
## 📋 PR #12004 — fix(platform): Fix Google Drive auto-credentials handling across the platform

**Author:** ntindle | **Files:** 12 changed (+1238/−45) | **Reviewed by:** 8-specialist automated review squad

---

### 🎯 Verdict: APPROVE WITH CONDITIONS

---

### What This PR Does

Separates "auto-credentials" (embedded in Google Drive file picker fields via `_credentials_id`) from regular user-provided credentials across the platform. Auto-credentials no longer appear in UI credential forms, CoPilot "missing credentials" prompts, or library preset guards. Forking an agent now clears embedded credential references to prevent cross-user credential leakage. Also fixes Firefox blocking Google API script loads via `referrerPolicy`.

---

### Specialist Findings

🛡️ **Security** ✅ — Credential ownership is properly user-scoped via `user_id` in all `acquire()` and `get_creds_by_id` calls. Fork clearing in `_reassign_ids()` correctly prevents cross-user credential reuse. No credential leakage paths identified. Two low-severity notes: partial lock cleanup on multi-field failure (mitigated by Redis TTL) and non-recursive dict clearing (sufficient for current `GoogleDriveFile` model).

🏗️ **Architecture** ✅ — Design is fundamentally sound. The `is_auto_credential`/`input_field_name` fields on `CredentialsFieldInfo` are generic enough to extend to other providers. Clean separation between regular and auto credential paths. Suggestions: extract `_credentials_id` as a named constant (used as magic string in 4+ locations), add explicit guard in `run_block.py` for auto-credential blocks until SECRT-1911 is addressed, and consider a shared state-parser for the 5-state dict semantics.

⚡ **Performance** ⚠️ — `regular_credentials_inputs` and `auto_credentials_inputs` are uncached `@property` methods that each independently call `aggregate_credentials_inputs()`. In a single request flow, this can run the full aggregation 2-3x. Additionally, `_validate_node_input_credentials` calls `get_integration_credentials_store()` and `get_creds_by_id` per auto-credential field (N+1 pattern). Current impact is low (only Google Drive auto-creds exist), but should be cached.

🧪 **Testing** ✅ — Excellent coverage with ~1,056 lines of new tests across 4 files. Tests are well-structured with ticket references (SECRT-1772, OPEN-2895), descriptive names, and realistic mocks. All tests confirmed running (no skip markers), CI passes on 3.11/3.12/3.13. Gaps: no test for multi-field lock stranding failure path, no test for `_credentials_id` key-missing scenario in validation, no test for `get_creds_by_id` exception path, no test for empty-string `_credentials_id`.

📖 **Quality** ⚠️ — Strong overall. Error messages are dynamic (provider pulled from config), user-friendly, and actionable. Test organization is excellent. One item: stale docstring in `match_user_credentials_to_graph()` still references `aggregate_credentials_inputs()` instead of `regular_credentials_inputs` (flagged by CodeRabbit, not yet fixed).

📦 **Product** ✅ — Complete fix for all primary user-facing issues. All 4 consumer paths (CoPilot prompts, credential input schema, library presets, execution mapping) updated to exclude auto-credentials. Fork clearing gives users clear, actionable error messages guiding re-authentication. Firefox `referrerPolicy` fix uses universal browser-supported values. Deferred `run_block.py` gap (SECRT-1911) is narrow and well-documented.

📬 **Discussion** ⚠️ — PR currently has **unresolved merge conflicts** with base branch (since Feb 17). majdyz's APPROVE was dismissed pending comment resolution — all substantive threads are now resolved (3 real bugs fixed: missing try/except, misplaced header, fork credential clearing changed from `=None` to `del`). Follow-up PR #12016 created for code duplication concern. 2 minor items unaddressed: stale docstring and credential ID in error message.

🔎 **QA** ✅ — Frontend loads correctly across all pages (landing, copilot, build, marketplace, library). Google Drive File Input Block renders on canvas. Backend healthy. `referrerPolicy` change properly implemented in `loadScript` helper. No PR-related console errors. Full Google OAuth flow untestable in local env (requires real credentials), but backend logic is thoroughly unit-tested.

**QA Screenshots:**
- [Landing](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-landing.png) | [After Signup](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-after-signup.png) | [Copilot](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-dashboard-copilot.png)
- [Build Page](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-build-page.png) | [Google Drive Block](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-google-drive-block.png)
- [Marketplace](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-marketplace.png) | [Library](https://pub-0d18968687014244a588926709ee135f.r2.dev/agent-browser/pr-12004-library.png)

---

### Conditions (must address before merge)

1. **Resolve merge conflicts** — PR has been in conflict since Feb 17. Must rebase/merge from base branch.
2. **Fix stale docstring** — `match_user_credentials_to_graph()` in `chat/tools/utils.py:233` still references `aggregate_credentials_inputs()`. Update to reference `regular_credentials_inputs`.

### Should Fix (follow-up OK)

1. **Cache `aggregate_credentials_inputs()`** — Use `@cached_property` or lazy init to avoid 2-3x redundant computation per request flow (`graph.py:619-637`).
2. **Extract `_credentials_id` as named constant** — Magic string used in 4+ locations across `manager.py`, `utils.py`, `graph.py`, `_base.py`. Define `AUTO_CREDENTIALS_ID_KEY = "_credentials_id"`.
3. **Hoist `get_integration_credentials_store()`** — Currently re-instantiated per iteration in validation loop (`utils.py:369`). Move outside the loop.
4. **Add `run_block.py` guard** — Explicit check that blocks with auto-credentials can't be run via `RunBlockTool` until SECRT-1911 is addressed. Currently would give confusing errors.
5. **Add edge-case tests** — Multi-field lock stranding, `_credentials_id` key-missing in validation, `get_creds_by_id` exception path, empty-string credential ID.

### Nice to Have

- Include credential ID in error messages for easier debugging (CodeRabbit suggestion)
- Consistent section header formatting in `graph_test.py`
- `auto_credentials_inputs` property is currently dead code — add a comment noting it's for future use or remove

---

### Risk Assessment

**Merge risk: LOW** — Well-tested (1056 lines of new tests, CI green), security model is sound, all existing tests pass. Changes are additive (new fields/properties) with minimal modification of existing logic.

**Rollback: EASY** — The new fields have defaults (`is_auto_credential=False`, `input_field_name=None`), so reverting wouldn't break existing data. The fork clearing and validation are new code paths that would simply be removed.

---

@ntindle This is a solid, well-tested security and UX fix. The credential separation design is clean and extensible. Main blockers are the merge conflicts and one stale docstring. Once those are resolved, this is good to merge. The should-fix items (caching, constant extraction, run_block guard) can follow up.

**Comment by ntindle** [comment]
Superseded by #12874 — merged into the combined autopilot beta blockers PR along with #12748 and #12588. Closing in favour of the combined PR.
